design/one-pager-controller-options-overrides.md

```diff
@@ -0,0 +1,42 @@
+# Controller Options Overrides
+
+- Owners:
+  - @max-melentyev
+- Reviewers: Crossplane Maintainers
+- Status: draft
+
+## Background
+
+There are cases when single configuration doesn't work for all controllers of a provider.
+Different Crossplane controllers may have very different load profile and number of managed resources.
+Consider examples in crossplane-contrib/provider-aws:
+
+- Reconciling route53 records is fast, but having many resources of this type may trigger AWS API
+throttling when all of them are reconciled with the same interval as other resources.
+- Reconciling EC2 instance is relatively slow, and having many managed instances can
+build up a long queue.
+
+Having a feature that makes it possible to override default poll interval for a specific controller
+would allow to avoid these issues without affecting other responsiveness of controllers in the provider.
+
+### Migrating Controllers to Event Driven Reconciliation
+
+Reconciling only updated resources would also help to resolve issues stated above.
+But this will require to change every controller to support it, and may also require
+additional infrastructure that will make it harder to use provider.
+
+Enabling this feature for selected controllers will require ability to increase poll interval
+for them, keeping it the same for other controllers.
+
+## Design
+
+### Providing Configuration
+
+Controller specific options can be provided with `--controller-option {controller-id}.{option}={value}`,
+for example `--controller-option ec2.instance.pollInterval=30m`.
```

Comment:
Another alternative could be more fine-grained controls via annotations on MR metadata, which would override the provider default. Per controller config you're proposing here is something in the middle; even though they have different purposes/use cases, not sure if we need both of these configuration knobs 🤔

Reply:
Sounds good, and it may require less changes in providers.
Though it makes resource owner responsible for configuring poll interval in the cases when it shouldn't care of it. For example, in provider-aws we implemented a monitor of AWS event that triggers reconciliation when resource is updated in AWS, and this allows us to increase poll interval for supported resources. Having this configuration as an annotation, will require updating all existing resources, and any CRD controllers that create these resources.

Reply:
I wonder if we'd have something like `spec.pollPolicy` (similar to `spec.managementPolicies`). I could imagine a few options:

* Whatever the provider default is
* A specific interval (30s, 2h, etc)
* Don't poll at all? This would basically disable drift correction - MRs would only be reconciled if they changed.